plugins/chain_plugin/chain_plugin.cpp

```diff
@@ -987,6 +989,52 @@ void chain_plugin::plugin_initialize(const variables_map& options) {
             }
          }
 
+      } else if (options.count( "json-snapshot" )) {
+         my->json_snapshot_path = options.at( "json-snapshot" ).as<bfs::path>();
+         EOS_ASSERT( fc::exists(*my->json_snapshot_path), plugin_config_exception,
+                     "Cannot load snapshot, ${name} does not exist", ("name", my->json_snapshot_path->generic_string()) );
+
+         // recover genesis information from the snapshot
+         // used for validation code below
+         auto infile = std::ifstream(my->json_snapshot_path->generic_string(), (std::ios::in));
+         istream_json_snapshot_reader reader(infile);
+         reader.validate();
+         chain_id = controller::extract_chain_id(reader);
+         infile.close();
+
+         EOS_ASSERT( options.count( "genesis-timestamp" ) == 0,
+                 plugin_config_exception,
+                 "--snapshot is incompatible with --genesis-timestamp as the snapshot contains genesis information");
```

Comment:
`--json-snapshot` not `--snapshot`. But more generally this looks like 40 lines copy pasted (almost?) verbatim. Is there a better way?